From patchwork Thu Dec 13 21:10:21 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10729919
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id D200F1759
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 13 Dec 2018 21:10:38 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id C57BA2C650
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 13 Dec 2018 21:10:38 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id B9E5E2C73F; Thu, 13 Dec 2018 21:10:38 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 689742C650
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 13 Dec 2018 21:10:38 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726533AbeLMVKh (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 13 Dec 2018 16:10:37 -0500
Received: from mail-pf1-f202.google.com ([209.85.210.202]:55648 "EHLO
        mail-pf1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726435AbeLMVKg (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 13 Dec 2018 16:10:36 -0500
Received: by mail-pf1-f202.google.com with SMTP id s71so2554627pfi.22
        for <kvm@vger.kernel.org>; Thu, 13 Dec 2018 13:10:36 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=1Wk780NEE7wrpsiZPoCc2skcF5oK0mGyYIa2/BCd8uE=;
        b=Amtg14Gic9uvtlpxfsz8jc2oRF6da95Zfwk7nE6h/KLBL1IXlquhRBwwNDh5f8ArIj
         IbkCNgILh/CReorIfQ77bNLCotzS6okWCkUqDVRtB6EIc+5p1upYipMw55isGH4kd9Ht
         u9XZZr/fFlEnekvRADSMNlC65Pm3hxDC9qRuqBW0J6lLvTRkbJ78Xq5qGqwgRsj+b5K2
         0l0bV/A72ME/TXWmQ2306UlvEUCQeQY0D6MDOUqp5yQfoeqr+LDrmCJhJTiZyi9E5j2g
         z7D+q/Qks7Nzp+QvjTcL22TbrNVrzldjkVz1+yfkEoxUN1kqXIhGarEesEF/MnT6L1kg
         V+pA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=1Wk780NEE7wrpsiZPoCc2skcF5oK0mGyYIa2/BCd8uE=;
        b=o+5tBSa9qVYtA18ZvNtFNb+92j/i6uB4rx11uiJnNSC5LdF2xlBLhnCgBKzL3YXt/x
         iEzFr8NnP60ER2YZZF+ARx727wPmn0EBYVyPpWbP9/WmpSQNNZXYqwZ29ZiuOrcSnQ3b
         h5/5luwl0qAWyIgL61mZ7CtusmJQHszfRX0Dn5Y5XOluZqEK9Zyoxpet2euWYv22+Jiw
         Fe0rLO4Iddjn5Q01HmTv0bLak+0i5jlCIy0SoTVjcuOktojnClND24f4dbUPbDMWaM2B
         QDz7PWcQjARf7CytBlJG32kTLLO4SSRw6i5EsapSNYLwe4bxLWuIZa2T6y+kp3SjIutD
         D/Mg==
X-Gm-Message-State: AA+aEWbj9hufn6iu6Ephe8xSXgf/MhFiA7uD7DUCwP5fKTBCWk8Uuf8H
        0TzsGGfYHCmau7WBrdxIdTYA6NQyVIvQKtvQ2y/4t6dFqTD/DerT1CzX0gnC1PiN/D0RJDMBWF/
        2sg/G4auir4592NW43dbiqZWtdunam776jF9U/T+khDwfQFvPpKXmXwt9YjlhUqQ=
X-Google-Smtp-Source: 
 AFSGD/U7J0iVWHhW6i8mZAp9NJFIC6ReYZNnPidj6s7smMMqL90rBIgpZMZB858+GoGSB91i5d6DQ3uO0ZwN3w==
X-Received: by 2002:a62:2258:: with SMTP id i85mr259874pfi.89.1544735435738;
 Thu, 13 Dec 2018 13:10:35 -0800 (PST)
Date: Thu, 13 Dec 2018 13:10:21 -0800
Message-Id: <20181213211021.202476-1-jmattson@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.20.0.rc2.403.gdbc3b29805-goog
Subject: [PATCH] kvm: Change offset in kvm_write_guest_offset_cached to
 unsigned
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>, Cfir Cohen <cfir@google.com>,
        Peter Shier <pshier@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Avoid potential out-of-bounds writes due to negative offsets. Note
that all current call sites are fine.

Fixes: 4ec6e8636256 ("kvm: Introduce kvm_write_guest_offset_cached()")
Reported-by: Cfir Cohen <cfir@google.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Cfir Cohen <cfir@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
Reviewed-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
Reviewed-by: Sean Christopherson <sean.j.christopherson@intel.com>
---
 include/linux/kvm_host.h | 3 ++-
 virt/kvm/kvm_main.c      | 3 ++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/include/linux/kvm_host.h b/include/linux/kvm_host.h
index c926698040e0d..a03d5e264e5e7 100644
--- a/include/linux/kvm_host.h
+++ b/include/linux/kvm_host.h
@@ -694,7 +694,8 @@ int kvm_write_guest(struct kvm *kvm, gpa_t gpa, const void *data,
 int kvm_write_guest_cached(struct kvm *kvm, struct gfn_to_hva_cache *ghc,
 			   void *data, unsigned long len);
 int kvm_write_guest_offset_cached(struct kvm *kvm, struct gfn_to_hva_cache *ghc,
-			   void *data, int offset, unsigned long len);
+				  void *data, unsigned int offset,
+				  unsigned long len);
 int kvm_gfn_to_hva_cache_init(struct kvm *kvm, struct gfn_to_hva_cache *ghc,
 			      gpa_t gpa, unsigned long len);
 int kvm_clear_guest_page(struct kvm *kvm, gfn_t gfn, int offset, int len);
diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 2679e476b6c39..065ee2fb40340 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -1965,7 +1965,8 @@ int kvm_gfn_to_hva_cache_init(struct kvm *kvm, struct gfn_to_hva_cache *ghc,
 EXPORT_SYMBOL_GPL(kvm_gfn_to_hva_cache_init);
 
 int kvm_write_guest_offset_cached(struct kvm *kvm, struct gfn_to_hva_cache *ghc,
-			   void *data, int offset, unsigned long len)
+				  void *data, unsigned int offset,
+				  unsigned long len)
 {
 	struct kvm_memslots *slots = kvm_memslots(kvm);
 	int r;
